src/SymmetricPositiveDefinite.jl

```diff
@@ -0,0 +1,482 @@
+
+@doc doc"""
+    exp!(M,y,x,v)
+
+compute the exponential map from `x` with tangent vector `v` on the [`SymmetricPositiveDefinite`](@ref)
+as a [`MetricManifold`](@ref) with [`LinearAffineMetric`](@ref) and modify `y`. The formula reads
+
+```math
+    \exp_x v = x^{\frac{1}{2}}\operatorname{Exp}(x^{-\frac{1}{2}} v x^{-\frac{1}{2}})x^{\frac{1}{2}},
+```
+where $\operatorname{Exp}$ denotes to the matrix exponential.
+"""
+function exp!(M::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric}, y, x, v) where N
+    e = eigen(Symmetric(x))
+    U = e.vectors
+    S = e.values
+    Ssqrt = Diagonal( sqrt.(S) )
+    SsqrtInv = Diagonal( 1 ./ sqrt.(S) )
+    xSqrt = Symmetric(U*Ssqrt*transpose(U))
+    xSqrtInv = Symmetric(U*SsqrtInv*transpose(U))
+    T = Symmetric(xSqrtInv * v * xSqrtInv)
+    eig1 = eigen( T ) # numerical stabilization
+    Se = Diagonal( exp.(eig1.values) )
+    Ue = eig1.vectors
+    xue = xSqrt*Ue
+    copyto!(y, xue*Se*transpose(xue) )
+    return y
+end
+
+@doc doc"""
+    exp!(M,y,x,v)
+
+compute the exponential map on the [`SymmetricPositiveDefinite`](@ref) `M` with
+[`LogCholeskyMetric`](@ref) from `x` into direction `v` and store the result in
+`y`. The formula reads
+
+````math
+\exp_x v = (\exp_l w)(\exp_l w)^\mathrm{T}
+````
+where $\exp_lw$ is the exponential map on [`CholeskySpace`](@ref), $l$ is the
+cholesky decomposition of $x$  and $w = l(l^{-1}vl^\mathrm{T}$.
+"""
+function exp!(M::MetricManifold{SymmetricPositiveDefinite{N},LogCholeskyMetric}, y, x, v) where N
+    (l,w) = spd_to_cholesky(x,v) 
+    exp!(CholeskySpace{N}(),y,l,w)
+    y .= y*y'
+    return y
+end
+
+@doc doc"""
+    log!(M,v,x,y)
+
+compute the logarithmic map at `x` to `y` on the [`SymmetricPositiveDefinite`](@ref)
+manifold with its default metric, [`LinearAffineMetric`](@ref) and modify `v`.
+"""
+log!(M::SymmetricPositiveDefinite{N}, v, x, y) where N = log!(MetricManifold(M,LinearAffineMetric()),v, x, y)
+
+@doc doc"""
+    log!(M,v,x,y)
+
+compute the exponential map from `x` to `y` on the [`SymmetricPositiveDefinite`](@ref)
+as a [`MetricManifold`](@ref) with [`LinearAffineMetric`](@ref) and modify `v`. The formula reads
+
+```math
+\log_x y = x^{\frac{1}{2}}\operatorname{Log}(x^{-\frac{1}{2}} y x^{-\frac{1}{2}})x^{\frac{1}{2}},
+```
+where $\operatorname{Log}$ denotes to the matrix logarithm.
+"""
+function log!(M::MetricManifold{SymmetricPositiveDefinite{N},LinearAffineMetric}, v, x, y) where N
+    e = eigen(Symmetric(x))
+    U = e.vectors
+    S = e.values
+    Ssqrt = Symmetric( Matrix( Diagonal( sqrt.(S) ) ) )
+    SsqrtInv = Symmetric( Matrix( Diagonal( 1 ./ sqrt.(S) ) ) )
+    xSqrt = Symmetric( U*Ssqrt*transpose(U) )
+    xSqrtInv = Symmetric( U*SsqrtInv*transpose(U) )
+    T = Symmetric( xSqrtInv * y * xSqrtInv )
+    e2 = eigen( T )
+    Se = Matrix( Diagonal( log.(max.(e2.values,eps()) ) ) )
+    Ue = e2.vectors
+    xue = xSqrt*Ue
+    copyto!(v, Symmetric(xue*Se*transpose(xue)))
```

Comment:
So basically I exchange the `copyto` with the `rmul!` and `mul!`? 

Interestingly – when I change as you propose here, the tests start to fail. I am getting numerically completely different results just changing the copy to the line you say. As I already said I am not an expert in these efficient methods, so I don't see why that happens....but I switched back and forth several times between the two lines of yours and my copy – any with your lines the tests fail.

Reply:
Yes, I see why. Did you make
```julia
Se = Diagonal( sqrt.(log.(max.(e2.values,eps())) ) )
xue = xSqrt*e2.vectors
rmul!(xue, Se)
mul!(v, xue, transpose(xue))
```
? The idea is to have the square root of your old `Se`, have it multiplied from the right to your old `xue`, thereby overwriting `xue`, and then multiply this "weighted" `xue` with its transpose. Then you should have the original `Se` in the middle, shouldn't you?

Reply:
Ah! Yes, we overwrote `xue` in the `rmul!`. But your new code has an additional sqrt basically, right? The problem (despite my question for efficiency of adding sqrt) is, that the log also returns negative values (which is completely correct) but we can not do sqrt of that...

Reply:
What of course _does_ work is (without sqrt of course) do `mul!(v,xue,Se*transpose(xue))`

Reply:
Ok, negative values are a show stopper. I don't think efficiency would be a problem otherwise. It's a simple function that you run on the diagonal vector. Doing `mul!(out, in, transpose(in))` is super-efficient. For BLAS floats, this is a specialized method in BLAS. I came across this recently; try to benchmark
```julia
A = rand(50, 20);
out = zeros(50, 50);
@benchmark mul!(out, A, transpose(A))
B = copy(transpose(A))
@benchmark mul!(out, A, B)
```
The result is the same, but different methods are running in the background. So I think, you may wish to bring things in the `mul!(out, A, A')` form whenever possible.

Reply:
Thanks for all the explanation – I'll keep in mind that `mul!(out, A, A')` is something to look for and use when possible.